README.md
# PyeDNA
![PyeDNA](https://github.com/drericstrong/pyedna/blob/master/images/pyedna_small.jpg)

[![PyPI version](https://badge.fury.io/py/pyedna.svg)](https://badge.fury.io/py/pyedna)
[![Documentation Status](https://readthedocs.org/projects/pyedna/badge/?version=latest)](http://pyedna.readthedocs.io/en/latest/?badge=latest)

PyeDNA ("pie-dee-en-ay") is a Python wrapper library for the C++ EzDnaApi,
written for data analysts who wish to work with eDNA data in the context of a
pandas DataFrame. By converting eDNA data into a DataFrame, data analysis can 
be accomplished using familiar tools like scikit-learn, statsmodels, etc. New
functions will be added upon request.

## Disclaimer
PyeDNA is a wrapper library for the API of a data historian called eDNA. eDNA 
is developed by InStepSoftware, LLC (http://www.instepsoftware.com/), 
who holds all rights to the eDNA software. PyeDNA does not contain any 
proprietary code, and is merely a wrapper for functions that must be obtained 
from a legal, licensed version of EzDnaApi.dll.

This is fan-supported project and is not affiliated in any way with InStepSoftware, LLC.
The maintainer enjoys working with eDNA and wishes them the best. :)

## Getting Started
PyeDNA currently requires that a legal, licensed version of the EzDnaApi be 
located in the following directory:

**C:\Program Files (x86)\eDNA\EzDnaApi64.dll**

Currently, only the 64-bit version of eDNA is supported. Python 2 is not supported.

**Required libraries**: numba, numpy, pandas

PyeDNA can be installed using pip:

**pip install pyedna**

## Documentation
Current documentation can be found [here](https://pyedna.readthedocs.io/en/latest/).

## Basic Examples
All of the core data pulling functions are located in the GetHist function, 
which will return a pandas DataFrame with the timestamp, value, and status 
columns. For example, the following code will pull snap data from TESTPNT1 
over a 30-second interval:

> import pyedna.ezdna as dna

> tag = "TESTSITE.TESTSERVICE.TESTPNT1"       # format site.service.tag

> start = "12/01/16 01:01:01"                 # format mm/dd/yy hh:mm:ss

> end = "01/03/17 01:01:01"                   # format mm/dd/yy hh:mm:ss

> period = "00:00:30"                         # format hh:mm:ss

> df = dna.GetHist(tag, start, end, period=period, mode="snap")

Raw data may be obtained from TESTPNT1 using:

> df = dna.GetHist(tag, start, end, mode="raw")

Other supported pull types include Average, Interpolated, Max, and Min. Please
refer to eDNA documentation for more description about these pull types.

Multiple tags can be pulled (in Raw mode) at the same time using:

> tags = ["TESTSITE.TESTSERVICE.TESTPNT1", "TESTSITE.TESTSERVICE.TESTPNT2",
          "TESTSITE.TESTSERVICE.TESTPNT3", "TESTSITE.TESTSERVICE.TESTPNT4"]
          
> df = dna.GetMultipleTags(tags, start, end)

A list of connected services may be obtained using GetServices:

> services = dna.GetServices()

A list of point information for a given service can be found using GetPoints:

> points = dna.GetPoints("TESTSITE.TESTSERVICE")


CHANGELOG.rst
===============
 Change Log
===============

Version 0.14
------------
- FEATURE- GetServices allows you to get all connected eDNA service information
- FEATURE- GetPoints allows you to get information about all points in a service
- FEATURE- Number of connected services are printed when library is imported
- Better error handling for eDNA connection drops
- Before a data pull, there is now error checking to see if a point exists
- GetMultipleTags no longer automatically resamples and forward-fills data. The user should be in control of this.

Version 0.15
------------
- FEATURE- In the pulling functions, you can now use the desc_as_label parameter to use the point description as the DataFrame column name.
- FEATURE- In the pulling functions, you can now specify a custom column label.
- Better handling of non-ASCII characters in descriptions and units
- GetRTFull never returned a point description- alternative written
- Improved handling of unicode errors- non-Unicode characters are now ignored
- Consistency between ezdna and serv file formatting and dll calls
- Beginnings of a unit test framework
- Miscellaneous code cleanup

Version 0.16
------------
- MAJOR- Refactoring of all GetHistX methods into GetHist. Please use the "mode" parameter to specify the type of history call. Old methods still available.
- New DEPENDENCY- Numba
- Significant speed increase due to JIT compilation
- Bugfix in __init__ header
- Project documentation

Version 0.17
------------
- Bugfix in GetHist related to "switch" statement
- Minor documentation fixes
- Mocking the dna_dll variable so that RTD documentation can be automatically created

Version 0.17
-------------
- Bugfix- minor issue with a duplicated last point in every data pull
- Bugfix where GetPoints and GetServices did not return the first entry (due to eDNA apparently treating these functions in a completely different way than the GetHist functions)
- Fixed issue in GetMultipleTags where the pandas function drop_duplicates() removed too many rows. Removed duplicate indices only, instead.

Version 0.18
-------------
- Updated logo
- New and improved documentation
- Import function updated
- Minor bugfixes

Version 1.01
-------------
- Re-released using the 1.X scheme to fix versioning control (developer mistake)
- No other major updates

examples/examples.py
# -*- coding: utf-8 -*-

import pandas as pd
import pyedna.ezdna as dna

# The following code will pull snap data from TESTPNT1 over a 30-second interval:

tag = "TESTSITE.TESTSERVICE.TESTPNT1"      # format site.service.tag
start = "12/01/16 01:01:01"                 # format mm/dd/yy hh:mm:ss
end = "01/03/17 01:01:01"                   # format mm/dd/yy hh:mm:ss
period = "00:00:30"                         # format hh:mm:ss
df = dna.GetHist(tag, start, end, period=period, mode="snap")

# Raw data may be obtained from TESTPNT1 using:

df2 = dna.GetHist(tag, start, end)

# Other supported pull types include Average, Interpolated, Max, and Min. Please
# refer to eDNA documentation for more description about these pull types.

# Multiple tags can be pulled (in Raw mode) at the same time using:

tags = ["TESTSITE.TESTSERVICE.TESTPNT1", "TESTSITE.TESTSERVICE.TESTPNT2",
        "TESTSITE.TESTSERVICE.TESTPNT3", "TESTSITE.TESTSERVICE.TESTPNT4"]
df3 = dna.GetMultipleTags(tags, start, end)

# A list of connected services may be obtained using GetServices:

services = dna.GetServices()

# A list of point information for a given service can be found using GetPoints:

points = dna.GetPoints("TESTSITE.TESTSERVICE")